activerecord/lib/active_record/locking/optimistic.rb

```diff
@@ -91,8 +91,7 @@ def _update_row(attribute_names, attempted_action = "update")
             locking_column = self.class.locking_column
             lock_attribute_was = @attributes[locking_column]
 
-            update_constraints = _primary_key_constraints_hash
-            update_constraints[locking_column] = _lock_value_for_database(locking_column)
+            update_constraints = _contraints_hash
```

Comment:
It's confusing to me that we're using the local `_constraints_hash` below and not the one from `Persistence`.

I wonder if there's a way to achieve this without changing the API (even if it's internal), so we don't have to do things like `base_constraints = super` and over-factor this code. -- IMO, a little duplication is fine here (in `Locking::Optimistic`) than to refactor early.

Reply:
@zzak thanks for the feedback.
I went with this approach because I thought it would provide a consistent style when I looked at how some of the other methods were implemented (such as `#_create_record` or `#_touch_row` being implemented in multiple modules in the chain, with each calling `super`. Which is the behaviour we wanted.
So now `#destroy_row` looks similar in both `CounterCache` and `Locking::Optimistic`. But you are right that this might be refactoring too early.

I do see how the use of `#_constraints_hash` could cause some confusion though. When looking in `Persistence` it isn't very clear that a few of those methods have been re-implemented in other modules.

---

One alternative I had thought of was to allow an optional constraints argument to the `#destroy_row` (and always pass constraints to `#_delete_row`), which could let keep only 1 implementation of `#_constraints_hash` in `Persistence` and the change in `Locking::Optimistic` would look like this:
```ruby
        # This extraction is just a personal preference, but it can still sit inside `#_update_row` and `#destroy_row`
        def _optimistic_locking_row_constraints
          base_constraints = _constraints_hash

          if locking_enabled?
            locking_column = self.class.locking_column
            base_constraints[locking_column] = _lock_value_for_database(locking_column)
          end

          base_constraints
        end

        def destroy_row(constraints = _optimistic_locking_row_constraints)
          # This could also be written as an early exit if that is a preference
          affected_rows = super

          if locking_enabled? && affected_rows != 1
            raise ActiveRecord::StaleObjectError.new(self, "destroy")
          end

          affected_rows
        end
```

Inside `Persistence` we could have:
```ruby

    def destroy_row(constraints = _constraints_hash)
      _delete_row(constraints)
    end

    def _delete_row(constraints)
      self.class._delete_record(constraints)
    end

```

What do you think of this approach? (I'm not against it, I just thought that it might seem like more of a change to the API)

---

NB: I just noticed a typo in my commit (`_contraints_hash` instead of `_constraints_hash`, missing first "s", so I'll need to push up that update anyway)

Reply:
@zzak Got any opinions on the above alternative approach?

Reply:
@choncou Is there a third option where we do it without changing the API? My guess is this would mean some duplication in the code (which I'm kind of fine with). Sorry for being so pedantic!

Reply:
@zzak I've pushed up a different approach. No changes in the method APIs, but this one relies on memoizing the number of affected rows. This way all `#destroy_row` functions always call `super` safely without risking `#_delete_row` being called multiple times, or with the wrong attributes.

What do we think of this ?